File: _static/code/perm_income/perm_inc_ir.py
r = 0.05
β = 1 / (1 + r)
T = 20  # Time horizon
S = 5   # Impulse date
σ1 = σ2 = 0.15


def time_path(permanent=False):
    "Time path of consumption and debt given shock sequence"
    w1 = np.zeros(T+1)
    w2 = np.zeros(T+1)
    b = np.zeros(T+1)
    c = np.zeros(T+1)
    if permanent:
        w1[S+1] = 1.0
    else:
        w2[S+1] = 1.0
    for t in range(1, T):
        b[t+1] = b[t] - σ2 * w2[t]
        c[t+1] = c[t] + σ1 * w1[t+1] + (1 - β) * σ2 * w2[t+1]
    return b, c


fig, axes = plt.subplots(2, 1, figsize=(10, 8))
p_args = {'lw': 2, 'alpha': 0.7}
titles = ['transitory', 'permanent']

L = 0.175

for ax, truefalse, title in zip(axes, (True, False), titles):
    b, c = time_path(permanent=truefalse)
    ax.set_title(f'Impulse reponse: {title} income shock')
    ax.plot(list(range(T+1)), c, 'g-', label="consumption", **p_args)
    ax.plot(list(range(T+1)), b, 'b-', label="debt", **p_args)
    ax.plot((S, S), (-L, L), 'k-', lw=0.5)
    ax.grid(alpha=0.5)
    ax.set(xlabel=r'Time', ylim=(-L, L))

axes[0].legend(loc='lower right')

plt.tight_layout()
plt.show()

File: _static/code/perm_income/perm_inc_ir.jl
const r = 0.05
const β = 1.0 / (1.0 + r)
const T = 20  # Time horizon
const S = 5   # Impulse date
const σ1 = 0.15
const σ2 = 0.15


function time_path(permanent=false)
    w1 = zeros(T+1)
    w2 = zeros(T+1)
    b = zeros(T+1)
    c = zeros(T+1)

    if permanent === false
        w2[S+2] = 1.0
    else
        w1[S+2] = 1.0
    end

    for t=2:T
        b[t+1] = b[t] - σ2 * w2[t]
        c[t+1] = c[t] + σ1 * w1[t+1] + (1 - β) * σ2 * w2[t+1]
    end

    return b, c
end


L = 0.175

b1, c1 = time_path(false)
b2, c2 = time_path(true)
p = plot(0:T, [c1 c2 b1 b2], layout=(2, 1),
         color=[:green :green :blue :blue],
         label=["consumption" "consumption" "debt" "debt"])
t = ["impulse-response, transitory income shock"
     "impulse-response, permanent income shock"]
plot!(title=reshape(t,1,length(t)), xlabel="Time", ylims=(-L, L), legend=[:topright :bottomright])
vline!([S S], color=:black, layout=(2, 1), label="")




